SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -0,0 +1,202 @@
+			if (Data.component_id == SpatialConstants::WORKER_COMPONENT_ID)
+			{
+				const Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+
+				const FString& WorkerType = SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_TYPE_ID);
+
+				// TODO(zoning): Currently, this only works if server workers never die. Once we want to support replacing
+				// workers, this will need to process UnassignWorker before processing AssignWorker.
+				if (WorkerType.Equals(SpatialConstants::DefaultServerWorkerType.ToString()) &&
+					!UnassignedVirtualWorkers.IsEmpty())
+				{
+					AssignWorker(SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_ID_ID));
+				}
+			}
+		}
+	}
+}
+
+// This will be called on the worker authoritative for the translation mapping to push the new version of the map
+// to the spatialOS storage.
+void SpatialVirtualWorkerTranslationManager::SendVirtualWorkerMappingUpdate()
+{
+	UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("SendVirtualWorkerMappingUpdate"));
+
+	// Construct the mapping update based on the local virtual worker to physical worker mapping.
+	Worker_ComponentUpdate Update = {};
+	Update.component_id = SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID;
+	Update.schema_type = Schema_CreateComponentUpdate();
+	Schema_Object* UpdateObject = Schema_GetComponentUpdateFields(Update.schema_type);
+
+	WriteMappingToSchema(UpdateObject);
+
+	check(Connection.IsValid());
+	Connection->SendComponentUpdate(SpatialConstants::INITIAL_VIRTUAL_WORKER_TRANSLATOR_ENTITY_ID, &Update);
+
+	// The Translator on the worker which hosts the manager won't get the component update notification,
+	// so send it across directly.
+	Translator->ApplyVirtualWorkerManagerData(UpdateObject);
+}
+
+void SpatialVirtualWorkerTranslationManager::QueryForWorkerEntities()
+{
+	UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("Sending query for WorkerEntities"));
+
+	checkf(!bWorkerEntityQueryInFlight, TEXT("Trying to query for worker entities while a previous query is still in flight!"));
+
+	// Create a query for all the system entities which represent workers. This will be used
+	// to find physical workers which the virtual workers will map to.
+	Worker_ComponentConstraint WorkerEntityComponentConstraint{};
+	WorkerEntityComponentConstraint.component_id = SpatialConstants::WORKER_COMPONENT_ID;
+
+	Worker_Constraint WorkerEntityConstraint{};
+	WorkerEntityConstraint.constraint_type = WORKER_CONSTRAINT_TYPE_COMPONENT;
+	WorkerEntityConstraint.constraint.component_constraint = WorkerEntityComponentConstraint;
+
+	Worker_EntityQuery WorkerEntityQuery{};
+	WorkerEntityQuery.constraint = WorkerEntityConstraint;
+	WorkerEntityQuery.result_type = WORKER_RESULT_TYPE_SNAPSHOT;
+
+	// Make the query.
+	Worker_RequestId RequestID;
+	check(Connection.IsValid());
+	RequestID = Connection->SendEntityQueryRequest(&WorkerEntityQuery);
```

Comment:
any reason not to move the declaration & initialisation to one line?

Reply:
Done.